### PR TITLE
Implement sparse semantics support in gradcheck (2nd try)

### PR DESCRIPTION
**Pull request by pearu**
Replaces https://github.com/pytorch/pytorch/pull/94714 that was reverted due to https://github.com/pytorch/pytorch/pull/94714#issuecomment-1442355648

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95405

cc @alexsamardzic @nikitaved @cpuhrsch @amjames @bhosmer @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

### CODE REVIEW

File: test/test_sparse.py [hunk 1]
```diff
@@ -402,7 +411,11 @@ def test_ctor_size_checks(self, device, dtype):
 
     @dtypes(*floating_and_complex_types_and(torch.float16, torch.bfloat16))
     @unittest.skipIf(TEST_WITH_CROSSREF, "generator unsupport triggers assertion error")
-    def test_to_dense(self, device, dtype):
+    @gradcheck_semantics()
+    @parametrize("fast_mode", [subtest(False, name='slow'), subtest(True, name='fast')])
+    def test_to_dense_with_gradcheck(self, device, dtype, gradcheck, fast_mode):
+        if not gradcheck.masked and fast_mode == 'slow':
+            self.skipTest(f'FIXME: to_dense_backward supports masked semantics only')
```
Comment (pearu) [inline test/test_sparse.py:418]: This is the fixing changeset relative to https://github.com/pytorch/pytorch/pull/94714

Comment (pearu) [reply test/test_sparse.py]: closing as out-of-date

**Review by albanD — APPROVED** [review]

**Comment by nikitaved** [comment]
Do we actually benefit from it in the long run? FWIW, https://github.com/pytorch/pytorch/issues/95550#issuecomment-1446045940.

**Comment by pearu** [comment]
> Do we actually benefit from it in the long run? FWIW, [#95550 (comment)](https://github.com/pytorch/pytorch/issues/95550#issuecomment-1446045940).

This PR fixes gradcheck for sparse inputs under non-masked semantics while keeping the masked semantics support.

So, I don't understand the question under the assumption that tensors with sparse layouts are semantically equivalent to strided tensors. Do you challenge this assumption?

**Comment by nikitaved** [comment]
@pearu, I would argue there is a better fix, namely, to_dense should not impose any masking, and that if the masking behavior is tested, it could be done in the forward function with the the use of `sparse_mask`. See examples in my point 4.

**Comment by pearu** [comment]
> Then I can test sparse semantics with:

This does not make sense to me. In my understanding, sparse semantics does not need to involve any masks.

Also, do you use `torch.autograd.gradcheck` from master (that computes numerical jacobians using masked semantics) or from this PR?

**Comment by nikitaved** [comment]
The mask is used to convert a dense tensor to its sparse representation. Which part do you not understand?

**Comment by pearu** [comment]
> 



> The mask is used to convert a dense tensor to its sparse representation. Which is confusing?

It is confusing for me because your examples are all about masked tensor applications.

Here we consider both non-masked and masked tensor applications. For instance,
```
torch.autograd.gradcheck(torch.Tensor.t, x)
```
should work for any tensor `x` regardless of its layout.

**Comment by nikitaved** [comment]
@pearu , I agree, it should work, but I am not convince that `masked=True/False` is the way to go as it is possible to circumvent its introduction.

**Comment by nikitaved** [comment]
OK, now I understand the confusion. The goal is to enable sparse tensor support with masked/non-masked semantics in gradcheck. Well, masked semantics in gradcheck is not necessarily the flaw, and could be used to test both, masked and non-masked semantics:
```python
>>> a_sparse = torch.tensor([[0, 1], [2, 3]], dtype=torch.float64).to_sparse().requires_grad_(True)
>>> a_nonsparse = a.detach().sparse_mask(torch.rand(2, 2).to_sparse()).requires_grad_(True)
>>> torch.autograd.gradcheck(lambda x: torch.Tensor.t(x).to_dense(),  (a_sparse,), check_sparse_nnz=True)
True  # Sparse semantics
>>> torch.autograd.gradcheck(lambda x: torch.Tensor.t(x).to_dense(),  (a_nonsparse,), check_sparse_nnz=True)
True # Non-Sparse semantics

```

**Comment by pearu** [comment]
@nikitaved using `randn` is likely equivalent to `ones` so that `a_nonmasked` is likely a full tensor. With full tensors, `gradcheck` result does not depend on chosen semantics.

**Comment by pearu** [comment]
From the discussions https://github.com/pytorch/pytorch/issues/95550#issuecomment-1446045940, above, and DM with @nikitaved , there exists another approach for fixing `gradcheck` without introducing the `masked` kw argument as explained below.

First, we'll postulate the following requirements:

1. Tensors using sparse layouts are semantically equivalent to tensors with strided layout

2. `gradcheck` analytical path should use a user-provided function and its inputs without any transformations to these to mimic the user application environment (recall that for the numerical path, the inputs must be densified so that perturbations can be applied to unspecified elements as well)

3. `gradcheck` operates under non-masked semantics only

The above should be sufficient for `gradcheck` to be always successful on a user-provided function that uses operations that backward functions implement non-masked semantics.

Now, if the user-specified operation uses operations that backward use masked semantics, the user must also specify the corresponding masks that are applied within the user-specified operation to inputs using `sparse_mask` or a similar tool.

**Comment by nikitaved** [comment]
@pearu , I am not sure that providing masks is a good strategy:

1. Under strided <-> sparse equivalence perturbations that result in a different mask might render masked functions non-differentiable. We have functions that are only differentiable in a rank-preserving neighborhood. For some reason, there is no option in gradcheck to preserve the rank, so I believe it is up to the user to generate a function that does what is expected, while preserving the general behavior of gradcheck no matter the space the inputs live in.

2. We have functions that are not masked in forward, but are in the backward, for example `torch.sparse.mm`. There is no way to test backward for these function inside gradcheck if `masked=True` is eliminated unless gradcheck performs a pre-processing of the inputs with `x = x.sparse_mask(x_mask)`, which contradicts the requirement of gradcheck to test the function provided with the inputs provided. This is to support point 1 in favor of giving all the freedom to the user.

**Comment by pearu** [comment]
@nikitaved 

> Under strided <-> sparse equivalence perturbations that result in a different mask might render masked functions non-differentiable

What do you mean by "perturbations that result in a different mask"? When masks are provided as inputs, these are typically bool tensors that are never perturbed. In general, only float or complex input tensors that have the `requires_grad` flag set to True, are perturbed, all other inputs are not.

I agree that the `masked` kw argument is not sufficient in general. Here's an example that fails for both `masked=True` and `masked=False` cases:
```python
>>> def foo(x):
...     y = torch.sparse.mm(x, x)
...     z = torch.mm(y, x)
...     return z
... 
>>> t = torch.tensor([[0, 1], [2, 3]], dtype=torch.float64).to_sparse().requires_grad_()
>>> torch.autograd.gradcheck(lambda x: foo(x).to_dense(), t, masked=False)
<snip>
torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor([[ 4.0000,  3.0000,  6.0000,  2.0000],
        [ 6.0000, 13.0000,  4.0000, 12.0000],
        [ 3.0000,  1.0000, 13.0000,  6.0000],
        [ 2.0000,  6.0000, 12.0000, 31.0000]], dtype=torch.float64)
analytical:tensor([[ 2.,  0.,  6.,  0.],
        [ 6., 13.,  4., 12.],
        [ 3.,  1., 13.,  6.],
        [ 2.,  6., 12., 31.]], dtype=torch.float64)

>>> torch.autograd.gradcheck(lambda x: foo(x).to_dense(), t, masked=True, check_sparse_nnz=True)
<snip>
torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor([[ 0.0000,  0.0000,  0.0000,  0.0000],
        [ 6.0000, 13.0000,  4.0000, 12.0000],
        [ 3.0000,  1.0000, 13.0000,  6.0000],
        [ 2.0000,  6.0000, 12.0000, 31.0000]], dtype=torch.float64)
analytical:tensor([[ 2.,  0.,  6.,  0.],
        [ 6., 13.,  4., 12.],
        [ 3.,  1., 13.,  6.],
        [ 2.,  6., 12., 31.]], dtype=torch.float64)
```
because `foo` uses functions that backward implementations assume different semantics.

Clearly, the above example makes sense only in masked semantics (because it uses operations that backwards are implemented for masked semantics). The numerical jacobian appears to be correct when the input mask is `[[False, True], [True, True]]` (this is implied from the fact that the implicit mask for torch.sparse.mm is the sparsity pattern of its input). To fix the analytical jacobian, the mask must be applied to the input when entering the function:
```python
>>> def foo(x, x_mask):
...     x = x.sparse_mask(x_mask)
...     y = torch.sparse.mm(x, x)
...     z = torch.mm(y, x)
...     return z
... 
>>> t_mask = torch.tensor([[0, 1], [1, 1]], dtype=torch.bool).to_sparse()
>>> torch.autograd.gradcheck(lambda *args: foo(*args).to_dense(), (t, t_mask), masked=True, check_sparse_nnz=True)
True
```

(Btw, here I am using pytorch version that suffers from issue https://github.com/pytorch/pytorch/issues/95550 . When applying the patch from https://github.com/pytorch/pytorch/issues/95550#issuecomment-1446678189, the example above still works.)

**Comment by nikitaved** [comment]
Oh, I thought the point was to manipulate mask inside the gradcheck. My bad.
Still the concern is there, what is a Masked semantics anyway?

**Comment by pearu** [comment]
> what is Masked semantics anyway?

Use https://pytorch.org/docs/master/masked.html as a starting point for the answer.

One can also ask "what is a tensor anyway?". A tensor is a set of values that are arranged over a regular grid where the grid nodes are indexed. The index-value pair is called a tensor element. One can define various high-level operations with tensors using operations defined on the tensor values and taking into account the regular arrangement of values on the grid.
A masked tensor is a generalization of the tensor where its values are arranged over an irregular grid. To represent a masked tensor, one can use a pair of regular tensors: values and mask, where mask values are 0 or 1 so that values elements are defined only when the corresponding mask element has value 1.
Clearly, not all high-level operations with tensors can be generalized for masked tensors in an unambiguous way but one can still try by introducing various restrictions. For instance, one could require that adding two masked tensors is defined only when the masked tensors have the same mask. Or more generally, one could define the addition of masked tensors formally using the operations defined on regular tensors: `(values1, mask1) + (values2, mask2) = (values1 + values2, mask1 * mask2)`.
Finally, to answer the original question, the Masked semantics is a study that defines the meaning of masked tensors and operations with masked tensors. Since we have regular tensors, then it is convenient to define the masked tensors and their operations in terms of regular tensors and operations with regular tensors as exemplified above.
That said, I don't think we have a clear understanding of what operations with masked tensors are needed in applications and what should be their definitions. For instance, take a matrix product and try to generalize this to masked tensors. There exist several definitions (e.g. (i) consider masked-out elements as 0 value, or (ii) consider masked-out elements as undefined, etc) but it is not clear which one will be useful for practical applications or which even make sense in terms of defining a well-defined algebra on masked tensors.

**Comment by nikitaved** [comment]
> > what is Masked semantics anyway?
> 
> Use https://pytorch.org/docs/master/masked.html as a starting point for the answer.
> 
> One can also ask "what is a tensor anyway?". A tensor is a set of values that are arranged over a regular grid where the grid nodes are indexed. 

Understanding of "what is a tensor" is important to understand what a sparse tensor is. Let me just gather some thoughts...

A real tensor, matrix for simplicity, of size `m x n` is isomorphic to `R^{n x m}` and is such a vector space with the canonical basis `{Delta_ij}_kl = 1 if i == k and j == l, 0 otherwise`. A sparse tensor is hence represented as `S = \sum_{i, j} s_ij * Delta_ij, where {i, j} \in S.indices(), and s_ij is the corresponding value` .

There is a clear bijection between a sparse matrix `S` and it's dense counterpart `S.to_dense()`. Hence, `to_dense` and `to_sparse` define isomorphisms between the vector space of dense matrices and the vector space of sparse matrices. As such, under the isomorphic map `to_sparse` we can use all the operations defined for dense tensors with sparse inputs, and we get the immediate relationship for maps that map dense inputs to dense inputs, namely
`f(d_1, ..., d_n) <=> f(d_1.to_sparse(), ..., d_n.to_sparse())`, this is how we get our behavior of sparse being just an optimization layout for functions defined over dense tensors.

Masked Semantics, or better be called Sparse Parametrization is more tricky when translating the mappings defined over dense inputs to sparse inputs, because dense mappings map full bases to full bases, not subsets of bases to subsets of bases. Recall the representation of a sparse tensor
`S = \sum_{i, j} s_ij * Delta_ij, where {i, j} \in S.indices(), and s_ij is the corresponding value`.
Now, any function `g(S)` that respects Masked Semantics actually maps the basis `Delta_ij, where {i, j} \in S.indices()`, so if we define `supp(S) := {Delta_ij | {i, j} \in S.indices()}`, then any perturbation of `S`, `dD` has to be in the span of `supp(S)`.
And this is what gradcheck is doing, it takes a permuation `dS \in span(supp(S)` and computes `g(S + dS)`, all fine here.
Since gradcheck actually tests `(to_dense o g)(S)`, and since `to_dense` does support Masked Semantics, we actually get that gradcheck tests a function `g` that maps a "masked tensor" to a "masked tensor", so that we get that
`dg(S) \ in span(supp(g(S)))`.

Disabling the masked semantics in `to_dense.backaward` will turn `to_dense` into a function that preserves the "natural" isomorphism between the dense and sparse tensors. This implies that we can still test the Masked Semantics grads with gradcheck, but not only do we need to mask the inputs, but also the outputs. Something along the lines of:

**Comment by pearu** [comment]
@nikitaved thanks for the formalized discussion! I think we are mostly on the same page but I have a few notes.

1. when we have agreed that regular tensors with strided and sparse layouts are isomorphic within non-masked or regular semantics of tensors then in the formal discussion we can just talk about tensors without referring to the used storage format. In fact, the masked tensor (defined as a pair of `values` and `mask` tensors) may have `values` and `mask` using both strided layout equally to using any sparse layout.

2. the ops `to_dense` and `to_sparse` are isomorphisms between the space of dense and sparse tensors under the implicit assumption that unspecified elements map to elements with zero value and elements with zero value map to unspecified elements, respectively. We should be very careful with this approach because all conclusions to be drawn will be valid under the assumption that unspecified elements are equivalent to zeros and vice-versa. In fact, an important point in masked semantics is that there exists no relation between unspecified elements (also called masked-out or ignored elements) and particular values because unspecified elements just do not exist (as the adjective "unspecified" also suggests) within masked semantics. If we decide to introduce the so-called fill value (be it 0 or some other finite or infinite value) then we are not in the domain of masked semantics anymore but in the domain of regular semantics.

   To put it differently, I hope we can overcome naturally relating sparse tensors with masked tensors to respect the fundamental assumption that sparse tensors (in terms of storage format) are regular tensors similar to strided tensors. If the mask and the indices set of a sparse tensor match then we just have a great optimization opportunity in implementing the operations and their backwards, but in general, the mask and the indices set of sparse tensors are unrelated.

3. "Do we agree that ... fixing both `to_dense` and setting `masked=False` to be the default behavior ?" - I totally agree. In fact, we have https://github.com/pytorch/pytorch/issues/95550 to fix `to_dense`, and `masked=False` is already the default for `gradcheck` due to https://github.com/pytorch/pytorch/pull/95405.

**Comment by nikitaved** [comment]
@pearu, thank you for your notes and this discussion!

Regarding 2: I believe we were careful in defining Sparse Parametrization. Even though, under zero nse, a sparse tensor is represented as `S = \sum_{i, j} s_ij * Delta_ij, where {i, j} \in S.indices(), and s_ij is the corresponding value`, in the mappings I discuss I consider not that we map this representation (just and embedding into dense tensors really), but the basis functions that span `S`, i.e `Delta_ij` and that values `s_ij`, these are the parameters. No other things exist outside of the support, `supp(S)`.
Now, when a `fill_value` is introduced, and if `S` is represented as
`S = \sum_{i, j in S.indices()} s_ij * Delta_ij + \sum_{i, j} fill_value * Delta_ij`, this is again just a representation, an embedding into a dense tensor.
And we have two options here as well: namely, is `fill_value` a parameter or not?
All-in-all, this part is not relevant to whether `to_sparse` or `to_dense` define an isomorphism. In this case they are just used as an embedding, not as an equivalence relationship.
